### PR TITLE
[WB-4888] (draft) expose filestream ratelimit errors to user + pass retry error messaging through to user process

### PR DESCRIPTION
**Pull request by stephchen**
https://wandb.atlassian.net/browse/WB-4888

Description
-----------
previously users who hit the filestream rate limits wouldn't see errors on the terminal (only logged to wandb/debug-internal.log). users who hit gql rate limits would see terminal errors logged from the background process; we decided it was undesirable to mix background/foreground logging. this PR exposes HTTP errors from filestream to the user and also moves HTTP error logging to the foreground:
- adds HttpResponse type to wandb_internal proto, with new repeated HttpResponse field within StatusResponse
- adds thread to RunStatusChecker to periodically poll for errors that prompted request retries in both internal_api and filestream
- SendManager maintains a retry_queue of HttpResponses on the background process filled during the retry loop and emptied in response to a poll from RunStatusChecker
- (most) terminal-logged outputs during retries are now redirected thru the retry_queue and logged from the user process instead

Testing
-------
ran jobs with a bunch of image logs under both gql and filestream rate limits, looking into unit/integration tests now.

output looks something like this:
```
Loss after 03072 examples: 0.370
Loss after 06272 examples: 0.250
wandb: 429 encountered (Filestream rate limit exceeded, retrying in 2.1993794997245546 seconds), retrying request
wandb: 429 encountered (Filestream rate limit exceeded, retrying in 4.521172651174481 seconds), retrying request
```

### CODE REVIEW

File: wandb/old/retry.py [hunk 1]
```diff
@@ -109,9 +112,13 @@ def __call__(self, *args, **kwargs):
...
-                    wandb.termlog(
-                        '{} ({}), entering retry loop. See {} for full traceback.'.format(
-                            self._error_prefix, e.__class__.__name__, util.get_log_file_path()))
+                    if isinstance(e, HTTPError):
+                        self.retry_callback(e.response.status_code, e.response.text)
+                    else:
+                        # todo: would like to catch other errors, eg wandb.errors.Error, ConnectionError etc
+                        # but some of these can be raised before the retry handler thread (RunStatusChecker) is
+                        # spawned in wandb_init
+                        wandb.termlog('{} ({}), entering retry loop.'.format(self._error_prefix, e.__class__.__name__))
```
Comment (stephchen) [inline wandb/old/retry.py:121]: eg when wandb can't connect to the server on startup we should print ConnectionErrors to terminal, but since RunStatusChecker only starts on run start, processing via retry_callback would result in hanging w/o any message

File: wandb/sdk/internal/file_stream.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def process_chunks(self, chunks):
                     util.to_human_size(util.MAX_LINE_SIZE),
                     util.to_human_size(len(chunk.data)),
                 )
-                wandb.termerror(msg, repeat=False)
+                wandb.termerror(msg, repeat=False)  # todo: don't want logging from background
```
Comment (stephchen) [inline wandb/sdk/internal/file_stream.py:48]: for completeness re: not logging from the background process we should be moving all of these to a similar framework

File: wandb/sdk/wandb_run.py [hunk 3]
```diff
@@ -147,14 +147,33 @@ class RunStatusChecker(object):
...
 
         self._join_event = threading.Event()
-        self._thread = threading.Thread(target=self.check_status)
-        self._thread.daemon = True
-        self._thread.start()
+        self._stop_thread = threading.Thread(target=self.check_status)
+        self._stop_thread.daemon = True
+        self._stop_thread.start()
+
+        self._retry_thread = threading.Thread(target=self.check_retries)
```
Comment (stephchen) [inline wandb/sdk/wandb_run.py:160]: i just added this stuff to RunStatusChecker, seems like it makes sense to have it all in one object though maybe needs a rename

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -147,14 +147,33 @@ class RunStatusChecker(object):
     For now, we just use this to figure out if the user has requested a stop.
     """
 
-    def __init__(self, interface: BackendSender, polling_interval: int = 15) -> None:
+    def __init__(self, interface: BackendSender, stop_polling_interval: int = 15, retry_polling_interval: int = 1) -> None:
```
Comment (stephchen) [inline wandb/sdk/wandb_run.py:150]: polling interval 1s is pretty arbitrary, i think we want something fairly frequent as error printing should feel close as possible to realtime

File: wandb/proto/wandb_internal.proto [hunk 5]
```diff
@@ -469,10 +469,17 @@ message GetSummaryResponse {
  */
 message StatusRequest {
   bool check_stop_req = 1;
+  bool check_retries = 2;
```
Comment (adrnswanberg) [inline wandb/proto/wandb_internal.proto:472]: I think this field name could be a little more generic. We could choose to return any kind of network status messages through here.

File: wandb/proto/wandb_internal.proto [hunk 6]
```diff
@@ -469,10 +469,17 @@ message GetSummaryResponse {
...
 }
 
 message StatusResponse {
   bool run_should_stop = 1;
+  repeated HttpResponse retry_responses = 2;
```
Comment (adrnswanberg) [inline wandb/proto/wandb_internal.proto:477]: same here

File: wandb/sdk/internal/sender.py [hunk 7]
```diff
@@ -105,9 +108,12 @@ def __init__(
         # State added when run_exit is complete
         self._exit_result = None
 
-        self._api = internal_api.Api(default_settings=settings)
+        self._api = internal_api.Api(default_settings=settings, retry_callback=self.retry_callback)
         self._api_settings = dict()
 
+        # queue filled by retry_callback
+        self._retry_q: "Queue[HttpResponse]" = multiprocessing.Queue()
```
Comment (adrnswanberg) [inline wandb/sdk/internal/sender.py:115]: nit: This probably isn't worth changing, but you could technically use a regular queue here.

**Review by adrnswanberg — CHANGES_REQUESTED** [review]
Looking good! Just a few nits. Also, would be good to add at least a functional test showing that errors are logged to the console. I think your py27 CI failures can be solved by just replacing fstrings with something str.format or the % syntax.

File: wandb/sdk/wandb_run.py (hunk 3 above)
Comment (raubitsj) [reply wandb/sdk/wandb_run.py:165]: can we just rename _stop_thread to _status_thread and do both stop checking and retry checking in a single thread

File: wandb/sdk/wandb_run.py [hunk 8]
```diff
@@ -147,14 +147,42 @@ class RunStatusChecker(object):
...
+        while not join_requested:
+            status_response = self._interface.communicate_status(
+                check_stop_req=False, check_retries=True
+            )
+            if status_response and status_response.retry_responses:
+                for hr in status_response.retry_responses:
+                    if (
+                        hr.http_status_code == 200 or hr.http_status_code == 0
+                    ):  # we use 0 for non-http errors (eg wandb errors)
+                        wandb.termlog(f"{hr.http_response_text}")
```
Comment (raubitsj) [inline wandb/sdk/wandb_run.py:180]: fstrings not allowed (not available in python3.5 and below)

this also might be a bit too verbose, doing a terminal print for every retry? seems like we should be tracking status instead about what type of errors are seen and printing a termlog every N seconds?

File: wandb/sdk/internal/sender.py [hunk 9]
```diff
@@ -716,7 +738,7 @@ def send_output(self, data):
             cur_time = time.time()
             timestamp = datetime.utcfromtimestamp(cur_time).isoformat() + " "
             prev_str = self._partial_output.get(stream, "")
-            line = u"{}{}{}{}".format(prepend, timestamp, prev_str, line)
+            line = "{}{}{}{}".format(prepend, timestamp, prev_str, line)
```
Comment (raubitsj) [inline wandb/sdk/internal/sender.py:741]: why the change?  i think the unicode thing was required for py27

File: wandb/old/retry.py (hunk 1 above)
Comment (raubitsj) [reply wandb/old/retry.py:121]: yep, We will fix this in another way, this is part of the message polling changes.

File: wandb/sdk/internal/file_stream.py [hunk 10]
```diff
@@ -45,7 +45,9 @@ def process_chunks(self, chunks):
                     util.to_human_size(util.MAX_LINE_SIZE),
                     util.to_human_size(len(chunk.data)),
                 )
-                wandb.termerror(msg, repeat=False)
+                wandb.termerror(
+                    msg, repeat=False
+                )  # todo: don't want logging from background
```
Comment (raubitsj) [inline wandb/sdk/internal/file_stream.py:50]: i assume these are here just for debug right now.  can they be removed now?

File: wandb/sdk/wandb_run.py (hunk 8 above)
Comment (raubitsj) [reply wandb/sdk/wandb_run.py:180]: Maybe have print all network messages as an internal setting so we can see all the warnings if we want to debug what is going on?  

Originally i figured that we would just be keeping track of stats in the sender process and this status check would just get these stats but the approach you did is a bit heavier weight but gives us better access to the actual errors.

File: wandb/sdk/wandb_run.py (hunk 3 above)
Comment (raubitsj) [reply wandb/sdk/wandb_run.py:165]: chatted with @stephchen, there are good reasons not to combine (status checker is doing a graphql request),  i agree this shouldnt be combined,  but since they are different they probably shouldnt share a message.  

Maybe your thread can be sync_status_thread, and the message called something like that SyncStatus,  or TransferStatus.   Having a thread that is getting this info is useful, and it is used for retry status in this implementation but it can easily pick up other info about how much has been synced or alert the user when the network sending threads are getting slow. etc

File: wandb/sdk/internal/sender.py (hunk 9 above)
Comment (stephchen) [reply wandb/sdk/internal/sender.py:741]: oops this might've gotten linted or something, reverting

File: wandb/sdk/wandb_run.py (hunk 8 above)
Comment (stephchen) [reply wandb/sdk/wandb_run.py:180]: so i actually didn't realize this at first but on line 113 of retry.py we only load the first retry failure (`self._num_iter == 2`) into the queue. in any case we only print at the same frequency as errors were already getting termlogged already before, there shouldn't be a visible change in frequency on the user end

**Review by adrnswanberg — APPROVED** [review]
awesome! looks great.
